aggregator_core/src/datastore.rs

```diff
@@ -61,6 +61,10 @@ pub mod test_util;
 
 // TODO(#196): retry network-related & other transient failures once we know what they look like
 
+/// List of schema versions that this version of Janus can safely run on. If any other schema
+/// version is seen, [`Datastore::new`] fails.
+const SUPPORTED_SCHEMA_VERSIONS: &[i64] = &[20230405185602, 20230417204528];
```

Comment:
It's easy to see when we'd add a new version to `SUPPORTED_SCHEMA_VERSIONS` (presumably, when we're creating a new DB migration, and we'd be reminded to do so because tests would begin failing until we do so).

But what's the process for removing a version from this list: how do we evaluate when it's correct to remove an item? more importantly, what reminds us to remove versions?

(somewhat separately from this, I wonder if it would be useful to have an env var or other configuration knob that controls which DB schema version is used for ephemeral datastores. this could be a useful tool, but not a complete solution, in evaluating if a given schema version still works with a given revision of the code.)

Reply:
> It's easy to see when we'd add a new version to `SUPPORTED_SCHEMA_VERSIONS` (presumably, when we're creating a new DB migration, and we'd be reminded to do so because tests would begin failing until we do so).
> 
> But what's the process for removing a version from this list: how do we evaluate when it's correct to remove an item? more importantly, what reminds us to remove versions?

With things as they stand in this PR, it's up to the person making a change to remove schema versions from the supported list. That does mean that nothing will remind us or force us to drop unsupported schemas. What would be really nice -- and I think this also speaks to your idea about configuring what schema version tests uses -- is for tests that involve the datastore to automatically be run against all the schema versions that are listed in this array. I think what we'd need is a macro for tests that use storage. The macro would generate code to instantiate an ephemeral datastore with the appropriate schema version, and it could also stamp out multiple tests, each one using a different supported version.

Then we'd get real assurance from the tests that Janus functions with all the schema versions we claim are supported, and it'd be completely obvious when an older schema is not supported: you'd see a plain test failure, and then could decide whether to drop that schema version or not.

Reply:
I'm experimenting with [`rstest`](https://docs.rs/rstest/latest/rstest/) for this, and it looks pretty promising!